File: src/commands/submit-claims-from-directory.ts
import * as commander from 'commander';
import { readdirSync, existsSync, mkdirSync, renameSync } from 'fs';
import { join, extname, basename } from 'path';
import chalk from 'chalk';

import { actionRunner, prompt } from '../utils.js';
import { getAccessToken } from '../config.js';
import { createClaim, getBenefitsWithCategories } from '../forma.js';
import { claimParamsToCreateClaimOptions } from '../claims.js';
import VERSION from '../version.js';
import { attemptToInferAllFromReceipt } from '../openai.js';

const command = new commander.Command();

interface Arguments {
  accessToken?: string;
  directory: string;
  processedDirectory?: string;
  openaiApiKey?: string;
  githubToken?: string;
}

// Supported file extensions for receipts
const SUPPORTED_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.pdf', '.heic'];

// Helper function to check if a file is a supported receipt type
const isSupportedReceiptFile = (filename: string): boolean => {
  const ext = extname(filename).toLowerCase();
  return SUPPORTED_EXTENSIONS.includes(ext);
};

// Helper function to get all receipt files from directory
const getReceiptFiles = (directory: string): string[] => {
  if (!existsSync(directory)) {
    throw new Error(`Directory '${directory}' does not exist.`);
  }

  try {
    const files = readdirSync(directory);
    return files.filter(isSupportedReceiptFile).map((file) => join(directory, file));
  } catch (error) {
    throw new Error(
      `Could not read directory '${directory}': ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
  }
};

// Helper function to move file to processed directory
const moveFileToProcessed = (sourceFile: string, processedDir: string): void => {
  try {
    if (!existsSync(processedDir)) {
      mkdirSync(processedDir, { recursive: true });
    }

    const filename = basename(sourceFile);
    const destinationFile = join(processedDir, filename);

    // If destination file already exists, add a timestamp suffix
    let finalDestination = destinationFile;
    if (existsSync(destinationFile)) {
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
      const ext = extname(filename);
      const nameWithoutExt = basename(filename, ext);
      finalDestination = join(processedDir, `${nameWithoutExt}-${timestamp}${ext}`);
    }

    renameSync(sourceFile, finalDestination);
    console.log(chalk.blue(`Moved processed receipt to: ${finalDestination}`));
  } catch (error) {
    console.error(
      chalk.red(
        `Warning: Could not move file ${sourceFile} to processed directory: ${
          error instanceof Error ? error.message : String(error)
        }`,
      ),
    );
    console.error(
      chalk.red('The claim was submitted successfully, but the file was not moved.'),
    );
  }
};

command
  .name('submit-claims-from-directory')
  .version(VERSION)
  .description('Submit claims from all receipt files in a directory')
  .requiredOption(
    '--directory <directory>',
    'The directory containing receipt files to process. Supported file types: JPEG, PNG, PDF, HEIC.',
  )
  .option(
    '--processed-directory <processed-directory>',
    'The directory to move successfully processed receipts to after processing (defaults to `processed/` subdirectory under the specified directory)',
  )
  .option('--access-token <access_token>', 'Access token used to authenticate with Forma')
  .option(
    '--openai-api-key <openai_token>',
    'An optional OpenAI API key used to infer claim details from receipt images. This can also be configured using the `OPENAI_API_KEY` environment variable. You must provide this or a --github-token.',
    process.env.OPENAI_API_KEY,
  )
  .option(
    '--github-token <github-token>',
    'An optional GitHub Token to use GitHub Models to infer claim details from receipt images. This can also be configured using the `GITHUB_TOKEN` environment variable. You must provide this or an --openai-api-key.',
    process.env.GITHUB_TOKEN,
  )
  .action(
    actionRunner(async (opts: Arguments) => {
      const { directory, openaiApiKey, githubToken } = opts;

      const accessToken = opts.accessToken ?? getAccessToken();

      if (!accessToken) {
        throw new Error(
          "You aren't logged in to Forma. Please run `npx formanator login` first.",
        );
      }

      if (!openaiApiKey && !githubToken) {
        throw new Error(
          'You must provide either an OpenAI API key (--openai-api-key) or GitHub token (--github-token) to infer claim details from receipts.',
        );
      }

      // Set up processed directory
      const processedDirectory = opts.processedDirectory ?? join(directory, 'processed');

      // Get all receipt files from the directory
      const receiptFiles = getReceiptFiles(directory);

      if (receiptFiles.length === 0) {
        console.log(
          chalk.yellow(`No supported receipt files found in directory: ${directory}`),
        );
        console.log(
          chalk.yellow(`Supported file types: ${SUPPORTED_EXTENSIONS.join(', ')}`),
        );
        return;
      }

      console.log(
        chalk.green(`Found ${receiptFiles.length} receipt file(s) to process:`),
      );
      receiptFiles.forEach((file, index) => {
        console.log(`  ${index + 1}. ${basename(file)}`);
      });
      console.log();

      // Get benefits and categories for inference
      const benefitsWithCategories = await getBenefitsWithCategories(accessToken);

      let processedCount = 0;
      let skippedCount = 0;

      // Process each receipt file
      for (const [index, receiptFile] of receiptFiles.entries()) {
        const filename = basename(receiptFile);
        console.log(
          chalk.cyan(
            `\n--- Processing receipt ${index + 1}/${receiptFiles.length}: ${filename} ---`,
          ),
        );

        try {
          // Infer claim details from receipt
          console.log('Analyzing receipt...');
          const inferredDetails = await attemptToInferAllFromReceipt({
            receiptPath: receiptFile,
            benefitsWithCategories,
            openaiApiKey,
            githubToken,
          });

          // Display inferred details to user
          console.log(chalk.green('\nInferred claim details:'));
          console.log(`  Amount: ${chalk.yellow(inferredDetails.amount)}`);
          console.log(`  Merchant: ${chalk.yellow(inferredDetails.merchant)}`);
          console.log(`  Purchase Date: ${chalk.yellow(inferredDetails.purchaseDate)}`);
          console.log(`  Description: ${chalk.yellow(inferredDetails.description)}`);
          console.log(`  Benefit: ${chalk.yellow(inferredDetails.benefit)}`);
          console.log(`  Category: ${chalk.yellow(inferredDetails.category)}`);

          // Prompt user for confirmation
          console.log(
            chalk.white(
              '\nDo you want to submit this claim? Enter Y to proceed or N to skip:',
            ),
          );
          const userResponse = prompt('> ').trim().toLowerCase();

          if (userResponse === 'y' || userResponse === 'yes') {
            // Submit the claim
            console.log('Submitting claim...');
            const createClaimOptions = await claimParamsToCreateClaimOptions(
              { ...inferredDetails, receiptPath: [receiptFile] },
              accessToken,
            );
            await createClaim(createClaimOptions);

            console.log(chalk.green(`✅ Claim submitted successfully for ${filename}`));

            // Move file to processed directory
            moveFileToProcessed(receiptFile, processedDirectory);
            processedCount++;
          } else {
            console.log(chalk.yellow(`Skipped ${filename}`));
            skippedCount++;
          }
        } catch (error) {
          console.error(
            chalk.red(
              `❌ Error processing ${filename}: ${error instanceof Error ? error.message : String(error)}`,
            ),
          );
          skippedCount++;
        }
      }

      // Summary
      console.log(chalk.green(`\n--- Summary ---`));
      console.log(`Processed successfully: ${chalk.green(processedCount)}`);
      console.log(`Skipped: ${chalk.yellow(skippedCount)}`);
      console.log(`Total files: ${receiptFiles.length}`);

      if (processedCount > 0) {
        console.log(chalk.blue(`Processed receipts moved to: ${processedDirectory}`));
      }
    }),
  );

export default command;

File: src/commands/submit_claims_from_directory.rs
use std::fs;
use std::io::{IsTerminal, Write};
use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use chrono::Utc;
use colored::Colorize;

use crate::claims::{ClaimInput, claim_input_to_create_options};
use crate::cli::SubmitClaimsFromDirectoryArgs;
use crate::config::resolve_access_token;
use crate::forma::{create_claim, get_benefits_with_categories};
use crate::llm::{infer_all_from_receipt, is_provider_error};
use crate::prompt::prompt;
use crate::verbose;

const SUPPORTED_EXTENSIONS: &[&str] = &["jpg", "jpeg", "png", "pdf", "heic"];

fn is_supported_receipt(path: &Path) -> bool {
    path.extension()
        .and_then(|e| e.to_str())
        .map(|e| {
            let lower = e.to_ascii_lowercase();
            SUPPORTED_EXTENSIONS.contains(&lower.as_str())
        })
        .unwrap_or(false)
}

fn list_receipt_files(directory: &Path) -> Result<Vec<PathBuf>> {
    if !directory.exists() {
        bail!("Directory '{}' does not exist.", directory.display());
    }
    let mut files = Vec::new();
    for entry in fs::read_dir(directory)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_file() && is_supported_receipt(&path) {
            files.push(path);
        }
    }
    files.sort();
    Ok(files)
}

const PROGRESS_WIDTH: usize = 30;

/// Draw a single-line progress bar, rewriting it in place. Falls back to one
/// plain line per receipt when stdout isn't a terminal, so piped output and CI
/// logs stay readable.
fn draw_progress(done: usize, total: usize, label: &str) {
    let mut stdout = std::io::stdout();
    if !stdout.is_terminal() {
        println!("Analyzing receipt {done}/{total}: {label}");
        return;
    }
    // \r returns to the start of the line and \x1b[K clears whatever the
    // previous, possibly longer, filename left behind.
    print!("\r\x1b[K{}", progress_line(done, total, label));
    let _ = stdout.flush();
}

fn progress_line(done: usize, total: usize, label: &str) -> String {
    let filled = (PROGRESS_WIDTH * done.min(total) / total.max(1)).min(PROGRESS_WIDTH);
    let bar = format!(
        "{}{}",
        "\u{2588}".repeat(filled),
        "\u{2591}".repeat(PROGRESS_WIDTH - filled)
    );
    format!("[{}] {done}/{total} {label}", bar.cyan())
}

fn finish_progress() {
    if std::io::stdout().is_terminal() {
        println!();
    }
}

fn move_to_processed(source: &Path, processed_dir: &Path) -> Result<()> {
    fs::create_dir_all(processed_dir)?;
    let filename = source
        .file_name()
        .ok_or_else(|| anyhow::anyhow!("Receipt path has no filename: {}", source.display()))?;
    let mut destination = processed_dir.join(filename);
    if destination.exists() {
        let timestamp = Utc::now().format("%Y-%m-%dT%H-%M-%S-%3f").to_string();
        let ext = source.extension().and_then(|s| s.to_str()).unwrap_or("");
        let stem = source
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("receipt");
        let new_name = if ext.is_empty() {
            format!("{stem}-{timestamp}")
        } else {
            format!("{stem}-{timestamp}.{ext}")
        };
        destination = processed_dir.join(new_name);
    }
    fs::rename(source, &destination)?;
    println!(
        "{}",
        format!("Moved processed receipt to: {}", destination.display()).blue()
    );
    Ok(())
}

pub fn run(args: SubmitClaimsFromDirectoryArgs) -> Result<()> {
    verbose::set(args.verbose);
    let access_token = resolve_access_token(args.access_token.as_deref())?;

    let processed_directory = args
        .processed_directory
        .clone()
        .unwrap_or_else(|| args.directory.join("processed"));

    let receipt_files = list_receipt_files(&args.directory)?;
    if receipt_files.is_empty() {
        println!(
            "{}",
            format!(
                "No supported receipt files found in directory: {}",
                args.directory.display()
            )
            .yellow()
        );
        println!(
            "{}",
            format!(
                "Supported file types: {}",
                SUPPORTED_EXTENSIONS
                    .iter()
                    .map(|e| format!(".{e}"))
                    .collect::<Vec<_>>()
                    .join(", ")
            )
            .yellow()
        );
        return Ok(());
    }

    println!(
        "{}",
        format!("Found {} receipt file(s) to process:", receipt_files.len()).green()
    );
    for (i, file) in receipt_files.iter().enumerate() {
        println!(
            "  {}. {}",
            i + 1,
            file.file_name().unwrap_or_default().to_string_lossy()
        );
    }
    println!();

    let benefits = get_benefits_with_categories(&access_token)?;

    // Every receipt is analysed up front so all the confirmation prompts
    // happen together, without waiting for the LLM in between.
    println!(
        "{}",
        format!(
            "Analyzing {} receipt(s) before review...",
            receipt_files.len()
        )
        .cyan()
    );
    let mut analyzed = Vec::new();
    for (index, receipt_file) in receipt_files.iter().enumerate() {
        draw_progress(
            index + 1,
            receipt_files.len(),
            &receipt_file
                .file_name()
                .unwrap_or_default()
                .to_string_lossy(),
        );
        let result = infer_all_from_receipt(
            receipt_file,
            &benefits,
            args.openai_api_key.as_deref(),
            args.openai_base_url.as_deref(),
            args.openai_model.as_deref(),
            args.copilot_cli_path.as_deref(),
        );
        // Nothing has been submitted yet, so a dead provider means bailing
        // out now rather than making the user sit through the rest.
        if let Err(error) = &result
            && is_provider_error(error)
        {
            finish_progress();
            return Err(result.unwrap_err()).context(
                "Aborted before reviewing any claims because the receipts could not be analysed",
            );
        }
        analyzed.push(result);
    }
    finish_progress();

    let mut processed = 0usize;
    let mut skipped = 0usize;

    for (index, (receipt_file, inferred)) in receipt_files.iter().zip(analyzed).enumerate() {
        let filename = receipt_file
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .into_owned();
        println!();
        println!(
            "{}",
            format!(
                "--- Processing receipt {}/{}: {filename} ---",
                index + 1,
                receipt_files.len()
            )
            .cyan()
        );

        let outcome = (|| -> Result<bool> {
            let inferred = inferred?;

            println!("{}", "\nInferred claim details:".green());
            println!("  Amount: {}", inferred.amount.yellow());
            println!("  Merchant: {}", inferred.merchant.yellow());
            println!("  Purchase Date: {}", inferred.purchase_date.yellow());
            println!("  Description: {}", inferred.description.yellow());
            println!("  Benefit: {}", inferred.benefit.yellow());
            println!("  Category: {}", inferred.category.yellow());

            println!(
                "\n{}",
                "Do you want to submit this claim? Enter Y to proceed or N to skip:".white()
            );
            let response = prompt("> ")?.trim().to_ascii_lowercase();
            if response == "y" || response == "yes" {
                println!("Submitting claim...");
                let claim = ClaimInput {
                    benefit: inferred.benefit,
                    category: inferred.category,
                    amount: inferred.amount,
                    merchant: inferred.merchant,
                    purchase_date: inferred.purchase_date,
                    description: inferred.description,
                    receipt_path: vec![receipt_file.clone()],
                };
                let opts = claim_input_to_create_options(&claim, &access_token)?;
                if args.dry_run {
                    println!("{}", "Dry run: skipping claim submission.".yellow());
                } else {
                    create_claim(&opts)?;
                }
                println!(
                    "{}",
                    format!("✅ Claim submitted successfully for {filename}").green()
                );
                if let Err(e) = move_to_processed(receipt_file, &processed_directory) {
                    eprintln!(
                        "{}",
                        format!(
                            "Warning: Could not move file {} to processed directory: {e}",
                            receipt_file.display()
                        )
                        .red()
                    );
                    eprintln!(
                        "{}",
                        "The claim was submitted successfully, but the file was not moved.".red()
                    );
                }
                Ok(true)
            } else {
                println!("{}", format!("Skipped {filename}").yellow());
                Ok(false)
            }
        })();

        match outcome {
            Ok(true) => processed += 1,
            Ok(false) => skipped += 1,
            Err(e) => {
                eprintln!("{}", format!("❌ Error processing {filename}: {e:#}").red());
                skipped += 1;
            }
        }
    }

    println!();
    println!("{}", "--- Summary ---".green());
    println!("Processed successfully: {}", processed.to_string().green());
    println!("Skipped: {}", skipped.to_string().yellow());
    println!("Total files: {}", receipt_files.len());
    if processed > 0 {
        println!(
            "{}",
            format!(
                "Processed receipts moved to: {}",
                processed_directory.display()
            )
            .blue()
        );
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::{PROGRESS_WIDTH, progress_line};

    fn blocks(line: &str) -> (usize, usize) {
        (
            line.matches('\u{2588}').count(),
            line.matches('\u{2591}').count(),
        )
    }

    #[test]
    fn progress_line_fills_in_proportion_and_never_overflows() {
        assert_eq!(blocks(&progress_line(0, 4, "a.jpg")), (0, PROGRESS_WIDTH));
        assert_eq!(
            blocks(&progress_line(2, 4, "a.jpg")),
            (PROGRESS_WIDTH / 2, PROGRESS_WIDTH / 2)
        );
        assert_eq!(blocks(&progress_line(4, 4, "a.jpg")), (PROGRESS_WIDTH, 0));
        // Degenerate inputs must not panic or overflow the bar.
        assert_eq!(blocks(&progress_line(0, 0, "a.jpg")), (0, PROGRESS_WIDTH));
        assert_eq!(blocks(&progress_line(9, 4, "a.jpg")), (PROGRESS_WIDTH, 0));
    }

    #[test]
    fn progress_line_shows_the_count_and_the_receipt_name() {
        let line = progress_line(3, 7, "receipt.jpg");
        assert!(line.contains("3/7"), "{line}");
        assert!(line.contains("receipt.jpg"), "{line}");
    }
}
